kong/runloop/handler.lua

```diff
@@ -1309,13 +1309,24 @@ return {
 
           local upstream_scheme = var.upstream_scheme
           if upstream_scheme == "http"  and balancer_data.port ~= 80 or
-             upstream_scheme == "https" and balancer_data.port ~= 443
+             upstream_scheme == "https" and balancer_data.port ~= 443 or
+             upstream_scheme == "grpc"  and balancer_data.port ~= 80 or
+             upstream_scheme == "grpcs" and balancer_data.port ~= 443
           then
             upstream_host = upstream_host .. ":" .. balancer_data.port
           end
 
           var.upstream_host = upstream_host
         end
+
+        -- the nginx grpc module does not offer a way to overrride
+        -- the :authority pseudo-header; use our internal API to
+        -- do so
+        local resty_grpc = require "resty.kong.grpc"
+        local ok, err = resty_grpc.set_authority(upstream_host)
```

Comment:
We should check to only call `set_authority` when the incoming request is indeed gRPC request. This call does involves a memory allocation that can not be freed immediately.